File: src/main/java/com/jlox/Lox.java
package com.jlox;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.nio.charset.Charset;
import java.nio.file.Files;
import java.nio.file.Paths;

public class Lox {
    public static boolean hadError;

    public static void main(String[] args) throws IOException {
        var len = args.length;
        if (len > 1) {
            System.out.println("Usage: jlox [path]");
        } else if (len == 1) {
            runFile(args[0]);
        } else {
            runPrompt();
        }
    }

    public static void runPrompt() throws IOException {
        InputStreamReader input = new InputStreamReader(System.in);
        BufferedReader reader = new BufferedReader(input);

        for (; ; ) {
            System.out.print("> ");
            var line = reader.readLine();
            // If line == null, user must have used CTRL+D
            if (line == null) break;
            run(line);
            hadError = false;
        }
    }

    private static void runFile(String path) throws IOException {
        byte[] bytes = new byte[0];
        try {
            bytes = Files.readAllBytes(Paths.get(path));
        } catch (java.nio.file.NoSuchFileException e) {
            System.err.println(e);
            System.exit(2);
        }
        run(new String(bytes, Charset.defaultCharset()));
        if (hadError) System.exit(1);
    }

    public static void run(String line) {
        System.out.println("Temp:" + line);
    }

    private static void report(int lineNumber, String where, String message) {
        System.err.println("[line " + lineNumber + "] Error" + where + ": " + message);
    }

    static void error(int lineNumber, String message) {
        report(lineNumber, "", message);
        hadError = true;
    }
}

File: src/main/scala/Lox.scala
package LoxApp

import scanner.Scanner
import tokens.{Token, TokenType}
import parser.{Parser, RuntimeError}
import runtime.Interpreter

import org.jline.reader.{LineReader, LineReaderBuilder}
import org.jline.terminal.TerminalBuilder

object LoxApp {
  private var hadError: Boolean = false
  private var hadRuntimeError: Boolean = false
  // Move Interpreter to top level scope so that it keeps state for REPL mode
  private val interpreter = Interpreter()

  def main(args: Array[String]): Unit = {
    val argLen = args.length
    if (argLen > 1) {
      println("Usage: slox [script]")
      sys.exit(64)
    } else if (argLen == 1) {
      runFile(args(0))
    } else {
      runPrompt()
    }
  }

  def runFile(path: String): Unit = {
    run(os.read(os.Path(path, os.pwd)))

    if (hadError) {
      sys.exit(65)
    } else if (hadRuntimeError) {
      sys.exit(70)
    }
  }

  def runPrompt(): Unit = {
    val terminal = TerminalBuilder
      .builder()
      .system(true) // Use system terminal
      .build()

    val reader: LineReader = LineReaderBuilder
      .builder()
      .terminal(terminal)
      .build()

    while (true) {
      val line = reader.readLine("> ")
      line match {
        case null   => return
        case "exit" => return
        case ""     => ()
        case _      => run(line)
      }
      hadError = false
    }
  }

  private def run(source: String): Unit = {
    val tokens = Scanner(source).scanTokens().toSeq
    val parser = Parser(tokens = tokens)
    val statements = parser.parse

    if (hadError) {
      ()
    } else {
      statements.foreach(stmt => interpreter.resolve(stmt))
      statements.foreach(stmt => interpreter.interpret(stmt))
      ()
    }
  }

  def error(line: Long, message: String): Unit = {
    report(line, "", message)
  }

  def error(token: Token, message: String): Unit = {
    if (token.tokenType == TokenType.EOF) {
      report(token.line, "at end", message)
    } else {
      report(token.line, s"at '${token.lexeme}'", message)
    }
  }

  def runtimeError(error: RuntimeError): Unit = {
    System.err.println(s"[line: ${error.token.line}]: ${error.message}")
    hadRuntimeError = true
  }

  def report(line: Long, where: String, message: String): Unit = {
    System.err.println(s"[line: ${line}] Error ${where}: ${message}")
    hadError = true
  }
}
